File: 2000-minimum-speed-to-arrive-on-time/minimum-speed-to-arrive-on-time.cpp
class Solution {
public:
    bool isValid(vector<int>& dist,double hour,int speed)
    {
        double time=0;   // calculating the amount of time taken for given speed.

        for(int i=0;i<dist.size()-1;i++)
        {
            time+=ceil((double)dist[i]/speed);  
        }
        time+= (double)dist[dist.size()-1]/speed;  // time taken to cover the last distance
        if(time>hour)      
        return false;
        return true;
    }

    int minSpeedOnTime(vector<int>& dist, double hour) 
    {
        int e=1e7;
        int s=1;
        int res=-1;

        while(s<=e)
        {
            int mid=s+(e-s)/2;
            if(isValid(dist,hour,mid)) // this speed is fine
            {
                res=mid;
                e=mid-1;          // checking if the task can be done in speed lesser than what we took right now
            }
            else
                s=mid+1;   // means we need to increase the speed
        }

        return res;
    }
};

File: 2000-minimum-speed-to-arrive-on-time/minimum-speed-to-arrive-on-time.java
class Solution {

    private boolean isValid(int[] dist, double speed, double timeToReach)
    {
       double time = 0;
       for(int i = 0; i < dist.length; i++)
       {
            time += (dist[i] / speed);
            if(time > timeToReach)
            return false;
            time = Math.ceil(time);
       }
       return true;
    }

    public int minSpeedOnTime(int[] dist, double hour) 
    {
        int end = 10000000;
        int start = 1;  

        int minSpeed = -1;
        while(start <= end)
        {
            int mid = start + (end-start)/2;
            // checking if this speed is fine
            // if its fine then try for lesser speeds
            if(isValid(dist, mid, hour))
            {
                minSpeed = mid;
                end = mid-1;
            }
            // means we need to increase speed
            else
            start = mid + 1;
        }  
        return minSpeed;
    }
}